[DEM-780] added docstring to ApiQI

QFer opened the pull request:
Added docstrings and revised docstring for all methods of Api.
Updated the unittests with some minor changes I made for getting consistency in output/interface etc.

quantumkoen commented:
@QFer awww code coverage dropped from 100% to 96.579%, any chance we get that back up to 100%?

QFer commented:
Added docstring. For some methods I added exception catching for raising an ApiError exception consistently. Unittests added for these exceptions. Furthermore type hints added for the whole file and removed types from docstrings.

quantumkoen commented on src/quantuminspire/api.py:
```diff
@@ -277,38 +419,70 @@ def _create_job(self, name, asset, project, number_of_shots, full_state_projecti
+                |                                   | its probability (float).
+                | histogram_url (str)               | Url to get the histogram with probabilities. This results in the
+                |                                   | OrderedDict as found in property histogram of result.
+                | measurement_mask (int)            | The measurement mask.
```
```suggestion
                | measurement_mask (int)            | (deprecated, unused) The measurement mask.
```

quantumkoen commented on src/quantuminspire/api.py:
```diff
@@ -277,38 +419,70 @@ def _create_job(self, name, asset, project, number_of_shots, full_state_projecti
+                |                                   | OrderedDict as found in property histogram of result.
+                | measurement_mask (int)            | The measurement mask.
+                | quantum_states_url (str)          | Url to get a list of quantum states.
+                | measurement_register_url (str)    | Url to get a list of measurement register.
```
```suggestion
                | measurement_register_url (str)    | Url to get a list of intermediate measurement register values.
```

quantumkoen commented on src/quantuminspire/api.py:
```diff
@@ -277,38 +419,70 @@ def _create_job(self, name, asset, project, number_of_shots, full_state_projecti
-        """ Gets the histogram results of the executed QASM code.
+        try:
+            result = self._action(['results', 'read'], params={'id': result_id})
+        except Exception as ex:
```
too broad

quantumkoen commented on src/quantuminspire/api.py:
```diff
@@ -277,38 +419,70 @@ def _create_job(self, name, asset, project, number_of_shots, full_state_projecti
+        try:
+            result = self._action(['results', 'read'], params={'id': result_id})
+        except Exception as ex:
+            raise ApiError('Result with id {} does not exist!'.format(result_id))
```
chain

quantumkoen commented on src/quantuminspire/api.py:
```diff
@@ -317,40 +491,61 @@ def get_raw_data(self, result_id):
-        """ Gets the properties of an asset.
+        try:
+            asset = self._action(['assets', 'read'], params={'id': asset_id})
+        except Exception as ex:
```
..

quantumkoen commented on src/quantuminspire/api.py:
```diff
@@ -317,40 +491,61 @@ def get_raw_data(self, result_id):
+        try:
+            asset = self._action(['assets', 'read'], params={'id': asset_id})
+        except Exception as ex:
+            raise ApiError('Asset with id {} does not exist!'.format(asset_id))
```
..

quantumkoen commented on src/quantuminspire/api.py:
```diff
@@ -362,46 +557,60 @@ def _create_asset(self, name, project, content):
         for attempt in attempts:
             time.sleep(sec_retry_delay)
-            status_message = '(id {}, iteration {})'.format(quantum_inspire_job.get_job_identifier(), attempt)
+            status_message = '(id: {}, iteration: {})'.format(quantum_inspire_job.get_job_identifier(), attempt)
```
It seems `status_message` is not used, can't we just remove it?

QFer replied:
Ok, i think it was there for debugging reasons, to get feedback while waiting for the job to complete.

quantumkoen commented on src/quantuminspire/api.py:
```diff
@@ -362,46 +557,60 @@ def _create_asset(self, name, project, content):
-            OrderedDict: The results of the executed qasm if succesfull else an empty dictionary if
-                         the results could not be collected.
+            The results of the executed cQASM if successful else an empty dictionary if
+            the results could not be collected within the given number of tries.
```
Refer to `get_result()` for a specification of the format of the result?

quantumkoen commented on src/quantuminspire/api.py:
```diff
@@ -410,31 +619,69 @@ def execute_qasm(self, qasm, backend_type=None, number_of_shots=256, collect_tri
+                           identifier: str = None, full_state_projection: bool = True,
+                           project: OrderedDict = None, job_name: str = None, user_data: str = '') -> QuantumInspireJob:
+        """ With this method a cQASM program (job) is scheduled to be executed asynchronously. The method returns
+            directly without waiting for the job to complete, as opposed to method 'execute_qasm' which waits for
```
```suggestion
            directly without waiting for the job to complete, as opposed to method `execute_qasm` which waits for
```

QFer replied:
Is there a convention for this?

quantumkoen replied:
In mark-down, back-tics are used to mark code and I always think of method names and such as code. It makes them stand out and most people will recognize it as something that refers to code.

quantumkoen commented on src/quantuminspire/api.py:
```diff
@@ -410,31 +619,69 @@ def execute_qasm(self, qasm, backend_type=None, number_of_shots=256, collect_tri
+            To execute a cQASM program a job is scheduled on a backend of type 'backend type' as given by the
+            parameter backend_type. Currently there are 3 backend types available:
+            1. 'QX single-node simulator'          (default for anonymous accounts)
+            2. 'QX single-node simulator SurfSara' (basic account credentials needed)
```
```suggestion
            2. 'QX single-node simulator SurfSara' (advanced account credentials needed)
```

quantumkoen commented:
Sorry, one more comment for the review: I'm getting a bunch of errors on the type hints from mypy:

```
(env) kfjmartens@ubuntu-vm:~/src/generic/SDK/src$ mypy --ignore-missing-imports -p quantuminspire
quantuminspire/api.py:124: error: "print" does not return a value
quantuminspire/api.py:261: error: Incompatible return value type (got "OrderedDict[Any, Any]", expected "List[OrderedDict[Any, Any]]")
quantuminspire/api.py:308: error: "print" does not return a value
quantuminspire/api.py:351: error: Incompatible return value type (got "OrderedDict[Any, Any]", expected "List[OrderedDict[Any, Any]]")
quantuminspire/api.py:370: error: Incompatible return value type (got "OrderedDict[Any, Any]", expected "List[OrderedDict[Any, Any]]")
quantuminspire/api.py:427: error: "print" does not return a value
quantuminspire/api.py:472: error: Incompatible return value type (got "OrderedDict[Any, Any]", expected "List[OrderedDict[Any, Any]]")
quantuminspire/api.py:489: error: Argument 1 to "get" of "QuantumInspireAPI" has incompatible type "Optional[Any]"; expected "str"
quantuminspire/api.py:490: error: Incompatible return value type (got "Union[OrderedDict[Any, Any], List[OrderedDict[Any, Any]], List[int]]", expected "List[int]")
quantuminspire/api.py:498: error: "print" does not return a value
quantuminspire/api.py:534: error: Incompatible return value type (got "OrderedDict[Any, Any]", expected "List[OrderedDict[Any, Any]]")
quantuminspire/api.py:620: error: Argument 2 to "_wait_for_completed_job" of "QuantumInspireAPI" has incompatible type "Optional[int]"; expected "int"
quantuminspire/api.py:690: error: Incompatible types in assignment (expression has type "UUID", variable has type "Optional[str]")
(env) kfjmartens@ubuntu-vm:~/src/generic/SDK/src$
```

QFer commented on src/quantuminspire/api.py:
```diff
@@ -48,97 +70,141 @@ def __init__(self, base_uri, authentication, project_name=None, coreapi_client_c
-    def get(self, uri_path):
-        """ Makes a request to get request with a client instance.
+    def get(self, uri_path: str) -> Union[OrderedDict, List[OrderedDict], List[int]]:
+        """ Method for making requests to the coreapi client instance to get some piece of information. The information
```
get is used more often, also in other classes. A new ticket is oke to fix it in the whole SDK.

QFer commented on src/tests/quantuminspire/test_api.py:
```diff
@@ -653,7 +723,7 @@ def __fake_no_results_job_handler(self, mock_api, document, keys, params=None, v
-                            ('queued_at', '2018-08-24T07,01,21.257557Z'),
+                            ('queued_at', '2018-08-24T07:01:21:257557Z'),
                             ('number_of_shots', 1)])
 
```
Ok

QFer commented on src/quantuminspire/api.py:
```diff
@@ -48,97 +70,141 @@ def __init__(self, base_uri, authentication, project_name=None, coreapi_client_c
-            uri_path (str): The URL where to request the data.
+            uri_path: The URL where to request the data.
+                      The uri_path has the form <api_uri><object><object-id><request>
+                      e.g. 'https://api.quantum-inspire.com/jobs/1234567/result/' to get the result of job with
```
I'll add a note for now: "This method will be made private in the near future. Usage is heavily discouraged."

QFer replied:
When making this method private, we must add all the necessary get_xxxxx methods that are needed.

QFer commented on src/quantuminspire/api.py:
```diff
@@ -48,97 +70,141 @@ def __init__(self, base_uri, authentication, project_name=None, coreapi_client_c
-        return self._action(['backendtypes', 'read'], params={'id': backend_id})
+        try:
+            backend_type = self._action(['backendtypes', 'read'], params={'id': backend_id})
+        except Exception as ex:
```
I'll change it to only catch ErrorMessage. This is thrown when the result contains an error.

QFer commented on src/quantuminspire/api.py:
```diff
@@ -48,97 +70,141 @@ def __init__(self, base_uri, authentication, project_name=None, coreapi_client_c
 
         Args:
-            backend_id (int): The backend identification number.
+            backend_id: The backend identification number.
```
Ok

QFer commented on src/quantuminspire/api.py:
```diff
@@ -48,97 +70,141 @@ def __init__(self, base_uri, authentication, project_name=None, coreapi_client_c
 
-    def get_default_backend_type(self):
-        """ Gets the default backend type.
+    def get_backend_types(self):
```
Ok

QFer commented on src/quantuminspire/api.py:
```diff
@@ -48,97 +70,141 @@ def __init__(self, base_uri, authentication, project_name=None, coreapi_client_c
+        try:
+            backend_type = self._action(['backendtypes', 'read'], params={'id': backend_id})
+        except Exception as ex:
+            raise ApiError('Backend type with id {} does not exist!'.format(backend_id))
```
Ok

QFer commented:
I handled all comments.
